CIP-optimized-uplc-serialization/README.md

```diff
@@ -0,0 +1,446 @@
+
+tags ```list``` and ```pair``` are the only tags that do require some other tag in order to be a defined type; since the twos always require some other type in order to be valid, the type application is implcit and it should be removed.
+
+during serialization this saves 4 bits for each ```list``` tag ( ```0111 0101``` becomes ```0101``` ) and 8 bits for pair tags which requires 2 type applications to be satisfied ( ```0111 0111 0110``` becomes ```0110``` ).
+
+two options are provided as alternative to the current design;
+
+option 1 helps backwards compatibitlity since it doesn't changes the serialization of same tags; but takes 4 bits per tag when 3 bits are enough to express all the alternatives. 
+
+this is why option 2 is present; allowing the 3 bits per tag serialization at the cost of missing bits for expansion in the evenience new constant types will be added; this cost needs to be discussed with the comunity since 1 bit per tag is a considerable improvment for complex types and is a 20% improvement on single element types. 
+
+### ```ByteString```s encoded as unsinged integer followed by the bytes
+
+ByteStrings are probably the values that have the most important impact on the serialized output;
+
+ByteStrings are used to represent a wide variety of types (```TokenName```, ```ValidatorHash```, ```PubKeyHash```, ```CurrencySymbol```, etc. are all ```ByteString``` wrappers ) ence widely used in smart contracts.
+
+citing the plutus core specificaiton draft released in june 2022 (latest specification at the time of writing)
+
+> Bytestrings are encoded by dividing them into nonempty blocks of up to 255 bytes and emitting each block in sequence.
+>
+> Each block is preceded by a single unsigned byte containing its length, and the end of the encoding is marked by a zero-length block (so the empty bytestring is encoded just as a zero-length block).
+>
+> Before emitting a bytestring, the preceding output is padded so that its length (in bits) is a multiple of 8;
+>
+> if this is already the case a single padding byte is still added; this ensures that contents of the bytestring are aligned to byte boundaries in the output.
+
+given the importance of bytestrings too much hoverhead is introduced; currently a ```ByteString``` serializes as
+
+```
+1) some padding that, depending on the context, may take up to 1 byte
+2) 1 (byte * chunk) indicating how many bytes are in the very next chunk
+3) 0 to 255 meaningful (bytes * chunk)
+4) 1 useless byte indicating no more chiunks are following
+```
+this implies ```(~1) + #chunks + 1``` meaningless bytes are added per ```ByteString```
+
+in the descripton above
+- step ```1``` allows for an easy serialization and deserialization but doesn't carries any meaningful informations; given the importance of ByteStrings it should be removed at the cost of an added shift while serializing/deserializing the value
```

Comment:
I think that might slow things down significantly, since I think we'd need to shift every byte in a bytestring if it didn't start at a byte boundary.  It's important that on-chain deserialisation be as fast as possible, so the extra expense might be problematic.  I'd like to see some benchmark results for this.

Reply:
It might be possible to implement deserialization without shifting bytes for efficiency

I'll leave here the [plu-ts implementation of the `readNBits`](https://github.com/HarmonicPool/plu-ts/blob/54332ac5716bcd0ddf51d512388a0cc74e904dde/src/onchain/UPLC/UPLCDecoder/index.ts#L142) function on which the deserialization process is based

I believe a very similar result can be achieved in Haskell using the [`Data.Binary.Get` monad](https://hackage.haskell.org/package/binary-0.10.0.0/docs/Data-Binary-Get.html)